README.md

```diff
@@ -1,2 +1,310 @@
-# validator
-A tiny runtime library for type assertions
+# Validator
+
+A tiny library for run time type assertions.
+
+## Installation
+
+```bash
+npm install @mapbox/validator
+```
+
+## Usage
+
+### Basic
+The basic usage 
+
+```javascript
+const v = require('@mapbox/validator');
+
+assertObj = v.assert(
+  v.shape({
+      names: v.arrayOf(v.string)
+  })
+);
+
+assertObj({names: ['ram', 'harry']}); // pass
+assertObj({names: ['john', 987]})// fail
+assertObj({names: 'john'})// fail
+```
+
+### Required
+
+Validation of each value is optional by default, which means passing `null` or `undefined` would not fail the assertion. You should wrap your validator with `v.required` in case you want strict assertion.
+
+```javascript
+assertObj = v.assert(
+  v.shape({
+      name: v.string
+  })
+);
+assertObj({}); // pass
+assertObj({name: undefined}); // pass
+assertObj({name: null}); // pass
+assertObj({name: 9}); // fail
+
+strictAssertObj = v.assert(
+     v.shape({
+      name: v.required(v.string)
+  })
+)
+strictAssertObj({}); // fail
+strictAssertObj({name: undefined}); // fail
+strictAssertObj({name: null}); // fail
+```
+
+### Composition
+You can compose any of the [Higher Order Validators](#higher-order-validators) to make complex validators.
+<details><summary>Example 1</summary>
+
+```javascript
+const personAssert = v.assert(
+  v.shape({
+    name: v.required(v.string),
+    family: v.arrayOf(
+      v.shape({
+        name: v.required(v.string),
+        relation: v.oneOf("wife", "husband", "son", "daughter")
+      })
+    )
+  })
+);
+
+// assertion passes
+personAssert({
+  name: "john",
+  family: [
+    {
+      name: "susy",
+      relation: "wife"
+    }
+  ]
+});
+
+// assertion fails
+personAssert({
+  name: "harry",
+  family: [
+    {
+      name: "john",
+      relation: "father" // family.0.relation must be a "wife", "husband", "son" or "daughter".
```

Comment:
Putting the comment may be a little confusing: I'd expect the command after the function call, since it's the output of the function.